File: src/lib/digger/policy.ts
const REFUSED_HOSTS = [
  "ecitizen.go.ke",
  "accounts.ecitizen.go.ke",
  "brs.go.ke",
  "itax.kra.go.ke",
  "odpc.go.ke",
  "linkedin.com",
  "www.linkedin.com",
  "facebook.com",
  "www.facebook.com",
  "instagram.com",
  "x.com",
  "twitter.com",
];

const LOGIN_PATH =
  /\/(login|signin|sign-in|signup|register|account|wp-admin|checkout|cart|paywall|captcha)(\/|$)/i;

const LOGIN_BODY =
  /recaptcha|hcaptcha|cf-turnstile|please (log|sign) in|password.*username|this content is for subscribers|paywall/i;

export function hostRefused(hostname: string) {
  const host = hostname.toLowerCase().replace(/^www\./, "");
  return REFUSED_HOSTS.some(
    (blocked) => host === blocked.replace(/^www\./, "") || host.endsWith(`.${blocked.replace(/^www\./, "")}`),
  );
}

export function urlRefused(raw: string): string | null {
  try {
    const url = new URL(raw);
    if (url.username || url.password) return "Credentials in the URL are refused.";
    if (hostRefused(url.hostname)) {
      return "That host is a login, registry, or social wall. VERIQ does not bypass it.";
    }
    if (LOGIN_PATH.test(url.pathname)) return "Login, checkout and admin paths are not crawled.";
    return null;
  } catch {
    return "URL is not valid.";
  }
}

export function bodyLooksGated(html: string, status: number) {
  if (status === 401 || status === 402 || status === 403) return "login_wall" as const;
  if (LOGIN_BODY.test(html.slice(0, 8000))) {
    if (/recaptcha|hcaptcha|cf-turnstile|captcha/i.test(html)) return "captcha" as const;
    if (/paywall|subscribers/i.test(html)) return "paywall" as const;
    return "login_wall" as const;
  }
  return null;
}

export function sameRegistrable(a: string, b: string) {
  const left = a.toLowerCase().replace(/^www\./, "");
  const right = b.toLowerCase().replace(/^www\./, "");
  return left === right;
}

File: tools/polite-crawler/crawler/policy.py
"""URL policy: public pages only. No login, registry, or social walls."""

from __future__ import annotations

from urllib.parse import urlparse

from crawler.errors import UrlRefused

REFUSED_HOST_SUFFIXES: tuple[str, ...] = (
    "ecitizen.go.ke",
    "brs.go.ke",
    "itax.kra.go.ke",
    "odpc.go.ke",
    "linkedin.com",
    "facebook.com",
    "instagram.com",
    "x.com",
    "twitter.com",
)

LOGIN_PATHS: tuple[str, ...] = (
    "/login",
    "/signin",
    "/sign-in",
    "/signup",
    "/register",
    "/account",
    "/wp-admin",
    "/checkout",
    "/cart",
    "/paywall",
    "/captcha",
)

LOGIN_BODY_MARKERS: tuple[str, ...] = (
    "recaptcha",
    "hcaptcha",
    "cf-turnstile",
    "please log in",
    "please sign in",
    "this content is for subscribers",
    "paywall",
)


def origin_of(url: str) -> str:
    """Return scheme://host[:port] for a URL."""
    parsed = urlparse(url)
    return f"{parsed.scheme}://{parsed.netloc}"


def host_refused(hostname: str) -> bool:
    """True if the host is a login, registry, or social wall."""
    host = hostname.lower().removeprefix("www.")
    return any(host == suffix or host.endswith(f".{suffix}") for suffix in REFUSED_HOST_SUFFIXES)


def looks_like_login_wall(html: str) -> bool:
    """True if the HTML looks like a login, CAPTCHA, or paywall."""
    sample = html[:8000].lower()
    return any(marker in sample for marker in LOGIN_BODY_MARKERS)


def assert_url_allowed(url: str) -> None:
    """Raise UrlRefused when the URL is out of policy."""
    parsed = urlparse(url)
    if parsed.scheme not in {"http", "https"}:
        raise UrlRefused("Only http(s) URLs are allowed.")
    if parsed.username or parsed.password:
        raise UrlRefused("Credentials in the URL are refused.")
    if host_refused(parsed.hostname or ""):
        raise UrlRefused("That host is a login, registry, or social wall. This collector does not fetch it.")
    path = (parsed.path or "/").lower()
    if any(path == prefix or path.startswith(prefix + "/") for prefix in LOGIN_PATHS):
        raise UrlRefused("Login, checkout, and admin paths are not crawled.")
